classes/Upgrader.php

```diff
@@ -85,138 +75,91 @@ public function __construct(string $version, FileLoader $fileLoader)
     {
-        if (empty($this->link)) {
-            $this->checkPSVersion();
+        if ($this->destinationRelease === null) {
+            $this->getOnlineDestinationRelease();
         }
 
         $destPath = realpath($dest) . DIRECTORY_SEPARATOR . $filename;
 
         try {
             $filesystem = new Filesystem();
-            $filesystem->copy($this->link, $destPath);
+            $filesystem->copy($this->destinationRelease->getZipDownloadUrl(), $destPath);
         } catch (IOException $e) {
             // If the Symfony filesystem failed, we can try with
             // the legacy method which uses curl.
-            Tools14::copy($this->link, $destPath);
+            Tools14::copy($this->destinationRelease->getZipDownloadUrl(), $destPath);
         }
 
         return is_file($destPath);
     }
 
+    /**
+     * @throws DistributionApiException
+     * @throws UpgradeException
+     */
     public function isLastVersion(): bool
     {
-        if (empty($this->link)) {
-            $this->checkPSVersion();
+        if ($this->getDestinationVersion() === null) {
+            return true;
         }
 
-        return !$this->need_upgrade;
+        return version_compare($this->currentPsVersion, $this->getDestinationVersion(), '>=');
     }
 
     /**
-     * checkPSVersion ask to prestashop.com if there is a new version. return an array if yes, false otherwise.
-     *
-     * @param bool $refresh if set to true, will force to download channel.xml
-     * @param string[] $array_no_major array of channels which will return only the immediate next version number
-     *
-     * @return array{'name':string,'link':string}|false
+     * @throws DistributionApiException
+     * @throws UpgradeException
      */
-    public function checkPSVersion(bool $refresh = false, array $array_no_major = ['minor'])
+    public function getOnlineDestinationRelease(): ?PrestashopRelease
     {
-        // if we use the autoupgrade process, we will never refresh it
-        // except if no check has been done before
-        $feed = $this->fileLoader->getXmlChannel($refresh);
-
-        // channel hierarchy :
-        // if you follow private, you follow stable release
-        // if you follow rc, you also follow stable
-        // if you follow beta, you also follow rc
-        // et caetera
-        $followed_channels = [];
-        $followed_channels[] = $this->channel;
-        switch ($this->channel) {
-        case 'alpha':
-            $followed_channels[] = 'beta';
-            // no break
-        case 'beta':
-            $followed_channels[] = 'rc';
-            // no break
-        case 'rc':
-            $followed_channels[] = 'stable';
-            // no break
-        case 'minor':
-        case 'major':
-        case 'private':
-            $followed_channels[] = 'stable';
+        if ($this->channel !== self::CHANNEL_ONLINE) {
+            throw new LogicException('channel must be online to retrieve the version dynamically');
```

Comment:
Not sure I got this exception right.
* The message does not seem to be related to the method contents
* What's preventing the Upgrader in archive mode to eventually look for updates?

Reply:
For me it is important to understand that this function is reserved for online mode, so that it is not accidentally called on a local channel. It's a kind of safeguard. 

If you are in an upgrade process in archive mode, you should not use this function, you will not search for your archive but the online version.

In which case are you in archive mode and you also want to have the online mode release?

Reply:
I would imagine both channels details when refreshing the update version selection after selecting the archive.
You display the local channel as selected, but you still need to display what would be the updated version via the only channel.

Reply:
The name of the method is enough for me indeed, and it is interesting to be able to compare the 2. I remove the condition